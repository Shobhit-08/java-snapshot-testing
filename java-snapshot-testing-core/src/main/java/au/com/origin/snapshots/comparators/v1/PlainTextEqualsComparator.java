package au.com.origin.snapshots.comparators.v1;

import au.com.origin.snapshots.Snapshot;
import au.com.origin.snapshots.comparators.SnapshotComparator;

public class PlainTextEqualsComparator implements SnapshotComparator {

  @Override
  public boolean matches(Snapshot previous, Snapshot current) {
    return previous.getBody().equals(current.getBody());
  }
}
