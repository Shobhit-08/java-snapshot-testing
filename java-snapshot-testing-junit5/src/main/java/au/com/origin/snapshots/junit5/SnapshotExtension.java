package au.com.origin.snapshots.junit5;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.SnapshotVerifier;
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig;
import au.com.origin.snapshots.config.SnapshotConfig;
import au.com.origin.snapshots.config.SnapshotConfigInjector;
import au.com.origin.snapshots.exceptions.SnapshotMatchException;
import au.com.origin.snapshots.logging.LoggingHelper;

import java.lang.reflect.Field;
import java.util.Arrays;

import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.extension.*;
import org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor;
import org.junit.jupiter.engine.descriptor.ClassTestDescriptor;

@Slf4j
public class SnapshotExtension
        implements AfterAllCallback,
        BeforeAllCallback,
        SnapshotConfigInjector,
        ParameterResolver,
        BeforeEachCallback {

    private SnapshotVerifier snapshotVerifier;

    @Override
    public void beforeAll(ExtensionContext context) {
        // don't fail if a test is run alone from the IDE for example
        boolean failOnOrphans = shouldFailOnOrphans();
        Class<?> testClass =
                context
                        .getTestClass()
                        .orElseThrow(() -> new SnapshotMatchException("Unable to locate Test class"));
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), testClass, failOnOrphans);
    }

    @Override
    public void afterAll(ExtensionContext context) {
        this.snapshotVerifier.validateSnapshots();
    }

    @Override
    public SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig();
    }

    private boolean shouldFailOnOrphans() {
        return false;
    }

    @Override
    public boolean supportsParameter(
            ParameterContext parameterContext, ExtensionContext extensionContext)
            throws ParameterResolutionException {
        boolean supports = parameterContext.getParameter().getType() == Expect.class;
        if (supports) {
            LoggingHelper.deprecatedV5(
                    log,
                    "Injecting 'Expect' via method a argument is no longer recommended. Consider using instance variable injection instead.");
        }
        return supports;
    }

    @Override
    public Object resolveParameter(
            ParameterContext parameterContext, ExtensionContext extensionContext)
            throws ParameterResolutionException {
        return new Expect(
                snapshotVerifier,
                extensionContext
                        .getTestMethod()
                        .orElseThrow(() -> new RuntimeException("getTestMethod() is missing")));
    }

    @Override
    public void beforeEach(ExtensionContext context) {
        if (context.getTestInstance().isPresent() && context.getTestMethod().isPresent()) {
            Arrays.stream(context.getTestClass().get().getDeclaredFields())
                    .filter(it -> it.getType() == Expect.class)
                    .findFirst()
                    .ifPresent(
                            field -> {
                                Expect expect = Expect.of(snapshotVerifier, context.getTestMethod().get());
                                field.setAccessible(true);
                                try {
                                    field.set(context.getTestInstance().get(), expect);
                                } catch (IllegalAccessException e) {
                                    throw new RuntimeException(e);
                                }
                            });
        }
    }
}
