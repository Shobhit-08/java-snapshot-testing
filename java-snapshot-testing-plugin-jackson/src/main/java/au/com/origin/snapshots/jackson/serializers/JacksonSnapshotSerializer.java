package au.com.origin.snapshots.jackson.serializers;

import au.com.origin.snapshots.logging.LoggingHelper;
import lombok.extern.slf4j.Slf4j;

@Deprecated
@Slf4j
public class JacksonSnapshotSerializer
    extends au.com.origin.snapshots.jackson.serializers.v1.JacksonSnapshotSerializer {

  public JacksonSnapshotSerializer() {
    super();
    LoggingHelper.deprecatedV5(
        log,
        "Update to `au.com.origin.snapshots.jackson.serializers.v1.JacksonSnapshotSerializer` in `snapshot.properties`");
  }
}
