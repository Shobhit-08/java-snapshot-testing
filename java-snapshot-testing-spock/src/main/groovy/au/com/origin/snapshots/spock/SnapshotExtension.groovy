package au.com.origin.snapshots.spock

import au.com.origin.snapshots.SnapshotVerifier
import au.com.origin.snapshots.config.PropertyResolvingSnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfig
import au.com.origin.snapshots.config.SnapshotConfigInjector
import org.spockframework.runtime.extension.AbstractAnnotationDrivenExtension
import org.spockframework.runtime.model.SpecInfo

class SnapshotExtension extends AbstractAnnotationDrivenExtension<EnableSnapshots> implements SnapshotConfigInjector {

    SnapshotVerifier snapshotVerifier;

    void visitSpecAnnotation(EnableSnapshots annotation, SpecInfo spec) {
        this.snapshotVerifier = new SnapshotVerifier(getSnapshotConfig(), spec.reflection, false)
    }

    void visitSpec(SpecInfo spec) {
        def snapshotMethodInterceptor = new SnapshotMethodInterceptor(snapshotVerifier)
        spec.allFeatures.featureMethod*.addInterceptor(snapshotMethodInterceptor)
        spec.addCleanupSpecInterceptor(snapshotMethodInterceptor)
    }

    @Override
    SnapshotConfig getSnapshotConfig() {
        return new PropertyResolvingSnapshotConfig()
    }
}