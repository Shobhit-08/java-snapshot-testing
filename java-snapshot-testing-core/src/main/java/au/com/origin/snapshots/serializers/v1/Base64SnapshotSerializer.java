package au.com.origin.snapshots.serializers.v1;

import au.com.origin.snapshots.Snapshot;
import au.com.origin.snapshots.SnapshotSerializerContext;
import au.com.origin.snapshots.serializers.SerializerType;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * This Serializer converts a byte[] into a base64 encoded string. If the input is not a byte[] it
 * will be converted using `.getBytes(StandardCharsets.UTF_8)` method
 */
public class Base64SnapshotSerializer implements SnapshotSerializer {
  private static final ToStringSnapshotSerializer toStringSnapshotSerializer =
      new ToStringSnapshotSerializer();

  @Override
  public Snapshot apply(Object object, SnapshotSerializerContext gen) {
    if (object == null) {
      toStringSnapshotSerializer.apply("", gen);
    }
    byte[] bytes =
        object instanceof byte[]
            ? (byte[]) object
            : object.toString().getBytes(StandardCharsets.UTF_8);
    String encoded = Base64.getEncoder().encodeToString(bytes);
    return toStringSnapshotSerializer.apply(encoded, gen);
  }

  @Override
  public String getOutputFormat() {
    return SerializerType.BASE64.name();
  }
}
