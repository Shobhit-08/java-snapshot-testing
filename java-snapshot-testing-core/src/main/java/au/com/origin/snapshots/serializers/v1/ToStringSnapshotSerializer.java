package au.com.origin.snapshots.serializers.v1;

import au.com.origin.snapshots.Snapshot;
import au.com.origin.snapshots.SnapshotFile;
import au.com.origin.snapshots.SnapshotSerializerContext;
import au.com.origin.snapshots.serializers.SerializerType;
import au.com.origin.snapshots.serializers.SnapshotSerializer;
import lombok.extern.slf4j.Slf4j;

/**
 * This Serializer does a snapshot of the {@link Object#toString()} method
 *
 * <p>Will render each toString() on a separate line
 */
@Slf4j
public class ToStringSnapshotSerializer implements SnapshotSerializer {

  @Override
  public Snapshot apply(Object object, SnapshotSerializerContext gen) {
    String body = "[\n" + object.toString() + "\n]";
    if (body.contains(SnapshotFile.SPLIT_STRING)) {
      log.warn(
          "Found 3 consecutive lines in your snapshot \\n\\n\\n. This sequence is reserved as the snapshot separator - replacing with \\n.\\n.\\n");
      body = body.replaceAll(SnapshotFile.SPLIT_STRING, "\n.\n.\n");
    }
    return gen.toSnapshot(body);
  }

  @Override
  public String getOutputFormat() {
    return SerializerType.TEXT.name();
  }
}
