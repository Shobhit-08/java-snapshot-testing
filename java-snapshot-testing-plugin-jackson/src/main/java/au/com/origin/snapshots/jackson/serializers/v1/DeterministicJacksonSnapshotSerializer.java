package au.com.origin.snapshots.jackson.serializers.v1;

import au.com.origin.snapshots.jackson.serializers.DeterministicCollectionModule;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Attempts to deterministically render a snapshot.
 *
 * <p>This can help in situations where collections are rendering in a different order on subsequent
 * runs.
 *
 * <p>Note that collections will be ordered which mar or may not be desirable given your use case.
 */
public class DeterministicJacksonSnapshotSerializer extends JacksonSnapshotSerializer {

  @Override
  public void configure(ObjectMapper objectMapper) {
    objectMapper.enable(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY);
    objectMapper.registerModule(new DeterministicCollectionModule());
  }
}
