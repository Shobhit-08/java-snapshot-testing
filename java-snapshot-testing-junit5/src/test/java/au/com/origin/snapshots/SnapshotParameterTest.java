package au.com.origin.snapshots;

import au.com.origin.snapshots.jackson.serializers.v1.JacksonSnapshotSerializer;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import java.util.stream.Stream;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

@ExtendWith({SnapshotExtension.class})
class SnapshotParameterTest {

  private Expect expect;

  static Stream<Arguments> testData() {

    return Stream.of(
        Arguments.of("Scenario2", "test input 1"),
        Arguments.of("Scenario2", "test input 1"),
        Arguments.of("Scenario2", "test input 1"),
        Arguments.of("Scenario3", "test input 2"),
        Arguments.of("Scenario3", "test input 2"));
  }

  @ParameterizedTest
  @MethodSource("au.com.origin.snapshots.SnapshotParameterTest#testData")
  void shouldSupportParameterizedTest(String scenario, String testInput, Expect expect) {
    expect.toMatchSnapshot("Duplicates are OK");
    expect.toMatchSnapshot("Duplicates are OK");
    expect.scenario("Scenario1").toMatchSnapshot("Additional snapshots need to include a scenario");
    expect
        .serializer(JacksonSnapshotSerializer.class)
        .scenario(scenario)
        .toMatchSnapshot(testInput);
  }

  @ParameterizedTest
  @MethodSource("au.com.origin.snapshots.SnapshotParameterTest#testData")
  void shouldSupportParameterizedTestViaInstanceVariable(String scenario, String testInput) {
    this.expect.toMatchSnapshot("Duplicates are OK");
    this.expect.toMatchSnapshot("Duplicates are OK");
    this.expect
        .scenario("Scenario1")
        .toMatchSnapshot("Additional snapshots need to include a scenario");
    this.expect
        .serializer(JacksonSnapshotSerializer.class)
        .scenario(scenario)
        .toMatchSnapshot(testInput);
  }
}
