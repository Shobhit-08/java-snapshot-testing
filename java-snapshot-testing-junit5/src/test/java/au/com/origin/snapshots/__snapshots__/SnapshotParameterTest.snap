au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest=[
Duplicates are OK
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTestViaInstanceVariable=[
Duplicates are OK
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTestViaInstanceVariable[Scenario1]=[
Additional snapshots need to include a scenario
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTestViaInstanceVariable[Scenario2]=[
  "test input 1"
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTestViaInstanceVariable[Scenario3]=[
  "test input 2"
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest[Scenario1]=[
Additional snapshots need to include a scenario
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest[Scenario2]=[
  "test input 1"
]


au.com.origin.snapshots.SnapshotParameterTest.shouldSupportParameterizedTest[Scenario3]=[
  "test input 2"
]